### DISCUSSION TITLE
about VRFCoordinatorV2Mock

### DISCUSSION THREAD

**Discussion by gen1ujs** [Cyfrin/foundry-full-course-cu#767] [Q&A] [answered]
```
contract AddConsumer is Script {
    function addConsumer(
        address raffle,
        address vrfCoordinator,
        uint64 subId
    ) public {
        console.log("Adding consumer contract", raffle);
        console.log("SubscriptionId", subId);
        console.log("chain id", block.chainid);

        vm.startBroadcast();
        VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, raffle);
        vm.stopBroadcast();
    }

    function addConsumerUsingConfig(address raffle) public {
        HelperConfig helperConfig = new HelperConfig();
        (, , address vrfCoordinator, , uint64 subId, , ) = helperConfig
            .activeNetwork();
        addConsumer(raffle, vrfCoordinator, subId);
    }

    function run() external {
        address raffle = DevOpsTools.get_most_recent_deployment(
            "Raffle",
            block.chainid
        );

        addConsumerUsingConfig(raffle);
    }
}
```

In Raffle project don't we use VRFCoordinatorV2Mock to make something on local chain? How it is able to do something on sepolia test net?

**Comment by gen1ujs** [comment]
I am asking that how VRFCoordinatorV2Mock is able to add my Raffle contract as a consumer on not a localchain?

**Reply by Niferu** [reply]
VRFCoordinatorV2Mock is not adding your contract on other than local network, it is done by already deployed and working contract on specific network.

Our script first recognize network and then decides if we need to deploy VRFCoordinatorV2 as it doesnt exists on local network or if we are using network like sepolia it is just picking up already deployed and exisitng contract address.

Hope that answer your question

**Reply by gen1ujs** [reply]
So in my code, should I do this :
```
 if (block.chainid == 31337) {
            vm.startBroadcast();
            VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, raffle);
            vm.stopBroadcast();
            console.log("localdaki kod");
        } else {
            vm.startBroadcast();
            VRFCoordinatorV2Interface(vrfCoordinator).addConsumer(
                subId,
                raffle
            );
            vm.stopBroadcast();
            console.log("interface kod");
        }
```

**Reply by gen1ujs** [reply]
But I don't understand this :

```
            vm.startBroadcast();
            VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, raffle);
            vm.stopBroadcast();
  
```
How this code actually add consumer to my vrf subscription?? I mean this code added my contract which is deployed to sepolia to this vrf subscription how this is possible if it is for localchain?

**Reply by Niferu** [reply, marked as answer]
@gen1ujs 

This contract did not added your subscription on sepolia network...

As I'm telling you, you have hardcoded vrfCoordinatorV2 contract address in your `HelperConfig.s.sol` and it is `0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625` and this contract did the job

This line `VRFCoordinatorV2Mock(vrfCoordinator).addConsumer(subId, raffle);` means that you are using some interface in this case VRFCoordinatorV2Mock, which is the same in sepolia, mainnet, local network etc, but once you pass this function argument, in this case contract address, so for local network some imaginary recently deployed one or for sepolia `0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625` this function will be executed by that passed address contract. You got it now?

**Reply by gen1ujs** [reply]
I get it. Thank you!

**Comment by Niferu** [comment]
Hello @gen1ujs 

There is contract already deployed and working with same interface and functions as our VRFCoordinatorV2Mock on sepolia test network and on mainnet etc. You can check proper contract addresses on ChainLink website. We are using VRFCoordinatorV2Mock only for local network as there is no such contract deployed because those networks are local working only for few minutes on our devices once we run some toolkits, that is why we need mock so copy of real working contracts from sepolia or mainnet or whatever thats what mocks are for :)
